### PR TITLE
Investigation into expanding use of preload & preconnect resource hints

### LINKED ISSUE #7144: Resource hints for media pages to improve performance
_Taken from [PR 7469](https://github.com/bbc/simorgh/pull/7469)_

## tl;dr - CONCLUSION

Following initial investigation, and conversations with @billthornton around [his own investgiations](https://paper.dropbox.com/doc/INNOVATION-DAY-Impact-of-preloading-resources--A50fusMCJFjn9tnkblbRLjgiAg-hhP8oDbUytXI3CGfCF21h) & @JibberJim - the conclusions can be summarised as:

- There are many reasons NOT to attempt to preconnect or preload 'real' media urls as detailed [in Jim's comment below](https://github.com/bbc/simorgh/pull/7469#issuecomment-676055798)
- Actual benefits of preconnecting to the bump/players domain are negligible due to a range of optimisations already in place by SMP. As detailed [in Jim's comment below](https://github.com/bbc/simorgh/pull/7469#issuecomment-676055798)
- Preloading bump _would_ work... but the question is; what is loading slower, while we preload bump?
- Ultimately the goal here was to try and allow SMP to have less to do (i.e. load quicker) **once initialised**. But, the bulk of our performance improvements can be achieved by getting SMP to **initialise quicker**, these are (as already discussed)

1. Stop using `polling.bbc.co.uk` domain on Canonical
2. Remove outer iframe on Canonical.

Pt 2. is a much larger discussion as we begin to maintain two media players, but pt 1. is fairly straightforward and I have a PR in progress to capture the work https://github.com/bbc/simorgh/pull/7581

----------

**Is your feature request related to a problem? Please describe.**
We should investigate if using resource hints on media pages could improve the speed of viewing content.

See here for more detail:
https://web.dev/preconnect-and-dns-prefetch/#streaming-media

There's a component in Simorgh that can easily do this for us if we decide it's worth the small amount of effort to implement.

https://github.com/bbc/simorgh/blob/latest/src/app/components/ResourceHints/index.jsx

We would drop this in at the page level.

> Only preconnect to critical domains you will use soon because the browser closes any connection that isn't used within 10 seconds.

This would be perfect for MAP and radio pages because we would expect the user to start playing media within 10 seconds of page load.

We should make sure that this works for iframe content resource loading. If not then we would need to implement in the news-av-player morph module

**Describe the solution you'd like**

**Describe alternatives you've considered**

**Testing notes**
[Tester to complete]

Dev insight: Will Cypress tests be required or are unit tests sufficient? Will there be any potential regression? etc

- [ ] This feature is expected to need manual testing.

**Additional context**

### PR DESCRIPTION
**Pull request by RichardPK**
Resolves #7144

## tl;dr - CONCLUSION

Following initial investigation, and conversations with @billthornton around [his own investgiations](https://paper.dropbox.com/doc/INNOVATION-DAY-Impact-of-preloading-resources--A50fusMCJFjn9tnkblbRLjgiAg-hhP8oDbUytXI3CGfCF21h) & @JibberJim - the conclusions can be summarised as:

- There are many reasons NOT to attempt to preconnect or preload 'real' media urls as detailed [in Jim's comment below](https://github.com/bbc/simorgh/pull/7469#issuecomment-676055798)
- Actual benefits of preconnecting to the bump/players domain are negligible due to a range of optimisations already in place by SMP. As detailed [in Jim's comment below](https://github.com/bbc/simorgh/pull/7469#issuecomment-676055798)
- Preloading bump _would_ work... but the question is; what is loading slower, while we preload bump?
- Ultimately the goal here was to try and allow SMP to have less to do (i.e. load quicker) **once initialised**. But, the bulk of our performance improvements can be achieved by getting SMP to **initialise quicker**, these are (as already discussed)

1. Stop using `polling.bbc.co.uk` domain on Canonical
2. Remove outer iframe on Canonical.

Pt 2. is a much larger discussion as we begin to maintain two media players, but pt 1. is fairly straightforward and I have a PR in progress to capture the work https://github.com/bbc/simorgh/pull/7581

------

## Investigation:
- It seems like principally we can benefit from two resource hints - `preload` for resources like media files/js scripts and `preconnect` to establish connections to origin domains that we will make calls to soon.

### Preface (updated 13/8/20)

The below tests were conducted with the assumption that Chrome's networking tools tell us everything about the connections being made. It turns out they probably don't. Comment below shows some stuff that Chrome wasn't reporting - which suggests that the conclusions below about preconnects/preloads **not** making a different, probably in fact **DO** make a difference.

https://github.com/bbc/simorgh/pull/7469#issuecomment-673505568

### Preconnect
We are already preconnecting to:

> https://cookie-oven.api.bbc.co.uk (`COOKIE_ORIGIN`)
> https://ichef.bbci.co.uk (`IMAGES_ORIGIN`)
> http://localhost:7080 / https://news.((test.))files.bbci.co.uk (`SIMORGH_PUBLIC_STATIC_ASSETS_ORIGIN`)
> https://logws1363.ati-host.net? / https://a1.api.bbc.co.uk/hit.xiti? (`SIMORGH_ATI_BASE_URL`)

_And if the service has custom fonts, i.e. Amharic_
> [ 'https://gel.files.bbci.co.uk', 'https://ws-downloads.files.bbci.co.uk' ]; (`FONTS_ORIGINS`)

In the context that any unused preconnect will be closed within 10s, additional preconnects **may** usefully be made to:

> isite?
> polling.((test.))bbc.co.uk
> emp.bbci.co.uk
> chartbeat?

I explored the impact of pre connecting on local to: `polling.test.bbc.co.uk`, `emp.stage.bbc.co.uk` & `emp.stage.bbci.co.uk` as various resources are fetched from these domains. There was no discernible benefit.

For example, taking the `polling.test.bbc.co.uk` endpoint 

**With** preconnect
with

**Without** preconnect
without

---

## Preload
Preload can be used to request scripts, documents, or images, before they are requested as part of the render. One potential use case for this would be SMP's `https://emp.stage.bbci.co.uk/emp/bump-4/bump-4.js` script

This **did** load the file a lot earlier. However, either because this file is later requested within the iframe, or because the request is coming from a different place, preloading it ends up requesting the file twice. The waterfall partly illustrates when the preloaded script comes in (much earlier) vs when it would regularly.

Screenshot 2020-08-12 at 15 58 06
---

Another potential use case here may be preloading the iframe itself, but this doesn't appear to have any impact on when the iframe is loaded. The fact nothing is preloaded suggests it could be an issue with the syntax, or with the polling endpoint not being recognised as a valid iframe?

**Without** iframe preloading:
Screenshot 2020-08-12 at 16 03 35

**With** iframe preload:
Screenshot 2020-08-12 at 16 03 04

This is even more evident when using the player with a placeholder. Even when preloading - the polling request isn't made until the placeholder is clicked (explicitly loading in the iframe). 

Lastly, I thought there may be a benefit to preload the placeholder image used by the player on some articles (note this is the `psammead-image-placeholder` used by articles - it isn't the SMP placeholder image seen on the regular player. )

But similarly, there appears to be no difference.

**With** placeholder image preloading:
with2

**Without** placeholder image preloading:
without2

## Conclusions

I've included my code in this PR so others can see for themselves if my conclusions above are fair, and that the code is valid. 

On the surface - it seems there's little to be gained here.

There **may** be a benefit to preconnecting to more services (i.e. the polling endpoint, isite) as part of our current ResourceHints within `/server/Document/component.jsx` but that is to be discussed.

There also **may** be benefit to preloading specific assets (video & audio) when we have the Simorgh > SMP hookup.

### CODE REVIEW

**Comment by jroebu14** [comment]
Thanks for looking into this, Richard.

I still think it's worth trying out as I mentioned possibly in the morph-module for CPS assets to preconnect to `https://emp.bbci.co.uk` since an example of a media file is `https://emp.bbci.co.uk/emp/media/blankv2.mp4`. The reason I think so is because it has been worth mentioning in https://web.dev/preconnect-and-dns-prefetch/#streaming-media I'm up for trying and finding a way to monitor if it has any effect.

> Another example where you may want to save some time in the connection phase, but not necessarily start retrieving content right away, is when streaming media from a different origin.

> Depending on how your page handles the streamed content, you may want to wait until your scripts have loaded and are ready to process the stream. Pre-connecting helps you cut the waiting time to a single round trip once you're ready to start fetching.

It might even be easier to ask the SMP team if they think preconnecting is a good idea.

**Comment by RichardPK** [comment]
Using the network debugging tool WebPageTest.org reveals a bit more than Chrome's network tab, so maybe preconnects are happening - I just haven't seen them reported in Google.

Currently - chrome’s network tab doesn’t show pre-connections being made to cookie-oven, ATI or ichef, (all of which are being preconnected to as part of the `assetOrigins` being passed into the `ResourceHints` component) but WebPageTest **does**.

This suggests that some of the preconnecting/preloading I've been doing may not have been captured by google, but IS happening. Frustratingly I can't run WebPageTest against local.

This is the report for https://www.bbc.com/pidgin/tori-51913962

https://www.webpagetest.org/result/200813_36_f10f414766da5d17ff97aa4d91d4cbcb/1/details/#waterfall_view_step1
Screenshot 2020-08-13 at 14 55 39

**Comment by JibberJim** [comment]
> I still think it's worth trying out as I mentioned possibly in the morph-module for CPS assets to preconnect to `https://emp.bbci.co.uk` since an example of a media file is `https://emp.bbci.co.uk/emp/media/blankv2.mp4`. 

So we would strongly not recommend pre-connecting for the media, you don't know enough of where the media is going to come from - the domains vary hugely depending on the type of media, the browser, media is distributed across 4 different CDNs with multiple host names on each, and they actually change quite often, preconnecting to all would be a disaster, and knowing which to preconnect to would be an impossible choice - you'd need to re-implement the entire complicated selection logic designed to ensure that we can support the volumes of data BBC has.   So you couldn't know which media domains to connect to.  The specific mp4 file you mention is not representative of the domains you use, it's specifically there just to work around the requirement that media playback starts with a user interaction.

Pre-connecting to the bump domain, and the players domains (emp.bbc.com or emp.bbc.co.uk depending on www tld) could be useful, however it's much less useful in reality due to http/2 coalescing, as the browser will almost certainly be reusing an existing TCP connection due to the many other resources which are shared on the akamai hosts.     We have telemetry on our iframe, the connection time, despite no-one pre-connecting it is already 80% of new connections that re-use another, and of course 95% of all loads are fully cached so no connection to the domain is required anyway.   For this reason, other than the bump domain, we do not recommend pre-connecting to our scripts.

We would like you to use the same gel font domains as us, that would help!

**Comment by jroebu14** [comment]
Thanks for the detailed explanation, Jim. I didn't realise that about the different domains. We thought this would be a quick win for a marginal gain but obviously this isn't the case. We're happy to take your advice on not pre-connecting.

> We would like you to use the same gel font domains as us, that would help!

I'll create an issue to look into this if there isn't one already.

**Comment by RichardPK** [comment]
See tl;dr in description 👍
